### ISSUE TITLE
PDM Lock File Procedure

### ISSUE THREAD

**Issue by mhogan-nwra** [pbrod/nvector#24]
# Introduction 

The PDM tool uses lock files to store the resolved project dependencies. The PDM tool has deprecated the cross-platform support of lock files. Given that this project uploads lock files, this issue provides a possible procedure to update lock files for the supported platforms.

# Lock File Updating

Naively running `pdm lock` will generate a cross-platform lock file. That file currently exists in the repo under the name `pdm.lock`. The PDM documentation now recommends to [specify the platform](https://pdm-project.org/latest/usage/lock-targets/#specify-lock-target-when-generating-lock-file). The project already has [two supporting PDM scripts](https://github.com/pbrod/nvector/blob/b7917df8e29f32fe0971025eb265fe4de603ba9d/pyproject.toml#L68) that generate lock files for the Windows and Linux platforms.

```language=shell
pdm run lock_win64 
pdm run lock_linux
```

# Recommended Changes

* Consider using the PDM scripts as a launching-off point to define cross-platform rules. I do not know how other projects resolve this issue. The current rule-set primarily resolves latest versions of dependencies instead of the minimum versions.
* Remove the cross-platform `pdm.lock` file in favor of platform-specific lock files. 
* Update the `CONTRIBUTING.rst` document with such rules.

**Comment by pbrod** [comment]
I added the pdm.lock file in order for the install command in python-package.yml to gain 10-30sec per combination of  platform and python version for installing the required the python packages.
Without the pdm.lock I would get a warning like this:

    Run pdm install -dG test
    WARNING: Lockfile does not exist

So if the recommendation is to have the lock files for each platform, then this should also be used  python-package.yml  file.

Do you have an idea on how to do this?

**Comment by mhogan-nwra** [comment]
I understand the concern about the efficiency gains. I think the required change to the `python-package.yml` would be to use the [following if-else case syntax supported by Actions](https://stackoverflow.com/a/57948488). 

```language=yaml
    steps:
    - uses: actions/checkout@v4
    - name: Set up PDM
      uses: pdm-project/setup-pdm@v4
      with:
        python-version: ${{ matrix.python-version }}
    - name: Install dependencies
      if: matrix.os == 'ubuntu-latest'
      run: |
        pdm install -dG test -L pdm_linux.lock
      if: matrix.os == 'windows-latest'
      run: |
        pdm install -dG test -L pdm_windows.lock
      if: matrix.os == 'macOS-latest'
      run: |
        pdm install -dG test -L pdm_macos.lock
    - name: Run Tests		
      run: |
         pdm run pytest
```

I could try this myself and submit a new pull request. Note that the `pdm_macos.lock` does not exist in the repo.

**Comment by pbrod** [comment]
Great! 👍

**Comment by pbrod** [comment]
Fixed in Platform Specific Lock Files #25
